test/distributed/fsdp/test_fsdp_meta.py

```diff
@@ -0,0 +1,315 @@
+# (c) Meta Platforms, Inc. and affiliates. Confidential and proprietary.
+
+import sys
+
+import torch
+import torch.distributed as dist
+import torch.nn as nn
+from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+from torch.distributed.fsdp.wrap import always_wrap_policy as always_wrap
+from torch.distributed.fsdp.wrap import wrap, enable_wrap
+from torch.testing._internal.common_fsdp import (
+    FSDPTest,
+)
+from torch.testing._internal.common_utils import (
+    TEST_WITH_DEV_DBG_ASAN,
+    run_tests,
+    parametrize,
+    instantiate_parametrized_tests,
+    sandcastle_skip_if,
+)
+from torch.testing._internal.common_distributed import (
+    skip_if_lt_x_gpu,
+)
+
+_TORCHDISTX_AVAIL = True
+try:
+    from torchdistx import deferred_init
+except ImportError:
+    _TORCHDISTX_AVAIL = False
+
+if not dist.is_available():
+    print("Distributed not available, skipping tests", file=sys.stderr)
+    sys.exit(0)
+
+if TEST_WITH_DEV_DBG_ASAN:
+    print(
+        "Skip dev-asan as torch + multiprocessing spawn have known issues",
+        file=sys.stderr,
+    )
+    sys.exit(0)
+
+
+class MyLinear(nn.Linear):
+    """
+    Linear layer with deterministic reset_parameters for testing.
+    """
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def reset_parameters(self, *args, **kwargs):
+        with torch.no_grad():
+            self.weight.fill_(1)
+
+class MyModel(nn.Module):
+    def __init__(self, device):
+        super().__init__()
+        self.lin1 = MyLinear(2, 2, bias=False, device=device)
+        self.lin2 = MyLinear(2, 2, bias=False, device=device)
+
+    def forward(self, x):
+        return self.lin2(self.lin1(x))
+
+    def reset_parameters(self, *args, **kwargs):
+        for m in [self.lin1, self.lin2]:
+            if not isinstance(m, FSDP):
+                m.reset_parameters()
+
+
+class NestedModel(nn.Module):
+    def __init__(self, device):
+        super().__init__()
+        torch.manual_seed(0)
+        torch.cuda.manual_seed(0)
```

Comment:
nit: It looks like `torch.cuda.manual_seed(0)` is covered by `torch.manual_seed(0)`.
https://discuss.pytorch.org/t/difference-between-torch-manual-seed-and-torch-cuda-manual-seed/13848/4

Reply:
Good point, actually the seeding can be removed here since param init is not random.